launcher/minecraft/MinecraftInstance.cpp

```diff
@@ -614,12 +618,41 @@ QProcessEnvironment MinecraftInstance::createLaunchEnvironment()
     }
 
     if (settings()->get("UseDiscreteGpu").toBool()) {
-        // Open Source Drivers
-        env.insert("DRI_PRIME", "1");
-        // Proprietary Nvidia Drivers
-        env.insert("__NV_PRIME_RENDER_OFFLOAD", "1");
-        env.insert("__VK_LAYER_NV_optimus", "NVIDIA_only");
-        env.insert("__GLX_VENDOR_LIBRARY_NAME", "nvidia");
+#ifdef WITH_QTDBUS
+        if (QDBusConnection::systemBus().isConnected()) {
+            QDBusInterface switcheroo("net.hadess.SwitcherooControl", "/net/hadess/SwitcherooControl", "org.freedesktop.DBus.Properties", QDBusConnection::systemBus());
+
+            if (switcheroo.isValid()) {
+                QDBusReply<QDBusVariant> reply = switcheroo.call(QStringLiteral("Get"), QStringLiteral("net.hadess.SwitcherooControl"), QStringLiteral("GPUs"));
+                if (reply.isValid()) {
+                    QDBusArgument arg = qvariant_cast<QDBusArgument>(reply.value().variant());
+                    QList<QVariantMap> gpus;
+                    arg >> gpus;
+
+                    for (const auto &gpu : gpus) {
+                        QString name = qvariant_cast<QString>(gpu[QStringLiteral("Name")]);
+                        bool defaultGpu = qvariant_cast<bool>(gpu[QStringLiteral("Default")]);
+                        if (!defaultGpu) {
+                            QStringList envList = qvariant_cast<QStringList>(gpu[QStringLiteral("Environment")]);                            
+                            for (int i = 0; i + 1 < envList.size(); i += 2) {
+                                env.insert(envList[i], envList[i + 1]);
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        else
+#endif
+        {
+            // Open Source Drivers
+            env.insert("DRI_PRIME", "1");
+            // Proprietary Nvidia Drivers
+            env.insert("__NV_PRIME_RENDER_OFFLOAD", "1");
+            env.insert("__VK_LAYER_NV_optimus", "NVIDIA_only");
+            env.insert("__GLX_VENDOR_LIBRARY_NAME", "nvidia");
+        }
```

Comment:
```suggestion
        bool envSet = false;
#ifdef WITH_QTDBUS
        if (QDBusConnection::systemBus().isConnected()) {
            QDBusInterface switcheroo("net.hadess.SwitcherooControl", "/net/hadess/SwitcherooControl", "org.freedesktop.DBus.Properties",
                                      QDBusConnection::systemBus());

            if (switcheroo.isValid()) {
                QDBusReply<QDBusVariant> reply =
                    switcheroo.call(QStringLiteral("Get"), QStringLiteral("net.hadess.SwitcherooControl"), QStringLiteral("GPUs"));
                if (reply.isValid()) {
                    QDBusArgument arg = qvariant_cast<QDBusArgument>(reply.value().variant());
                    QList<QVariantMap> gpus;
                    arg >> gpus;

                    for (const auto& gpu : gpus) {
                        QString name = qvariant_cast<QString>(gpu[QStringLiteral("Name")]);
                        bool defaultGpu = qvariant_cast<bool>(gpu[QStringLiteral("Default")]);
                        if (!defaultGpu) {
                            QStringList envList = qvariant_cast<QStringList>(gpu[QStringLiteral("Environment")]);
                            for (int i = 0; i + 1 < envList.size(); i += 2) {
                                env.insert(envList[i], envList[i + 1]);
                            }
                            envSet = true;
                            break;
                        }
                    }
                }
            }
        }
#endif
        if (!envSet) {
            // Open Source Drivers
            env.insert("DRI_PRIME", "1");
            // Proprietary Nvidia Drivers
            env.insert("__NV_PRIME_RENDER_OFFLOAD", "1");
            env.insert("__VK_LAYER_NV_optimus", "NVIDIA_only");
            env.insert("__GLX_VENDOR_LIBRARY_NAME", "nvidia");
        }
    }
```

Reply:
The original code was a bit ugly(else split by endif) but also did not work:
the dbus may be connected but the issue remained that the SwitcherooControl was not installed so it would still fail to fallback because the response form dbus is invalid

You may want to move the envSet=true; line a bit above: somewhere after `if (reply.isValid()) {` line so that if there was no dedicated GPU detected to not fall back

Reply:
code looks good enough to me and is nothing more but a stop-gap fix until a proper solution is made